tilings/algorithms/gridded_perm_reduction.py

```diff
@@ -0,0 +1,332 @@
+from collections import defaultdict
+from functools import partial
+from typing import Callable, Dict, Iterable, List, Optional, Set, Tuple
+
+from comb_spec_searcher.utils import cssmethodtimer
+
+from ..griddedperm import GriddedPerm
+
+Cell = Tuple[int, int]
+Requirement = Tuple[GriddedPerm, ...]
+
+func_calls: Dict[str, int] = defaultdict(int)
+func_times: Dict[str, float] = defaultdict(float)
+
+
+class GriddedPermReduction:
+    func_calls = func_calls
+    func_times = func_times
+    passes: Dict[int, int] = defaultdict(int)
+
+    @cssmethodtimer("GriddedPermReduction.__init__")
+    def __init__(
+        self,
+        obstructions: Tuple[GriddedPerm, ...],
+        requirements: Tuple[Tuple[GriddedPerm, ...], ...],
+        sorted_input: bool = False,
+    ):
+        # Only using MGP for typing purposes.
+        if sorted_input:
+            self._obstructions = obstructions
+            self._requirements = requirements
+        else:
+            self._obstructions = tuple(sorted(obstructions))
+            self._requirements = tuple(sorted(tuple(sorted(r)) for r in requirements))
+
+        self._minimize_griddedperms()
+
+    @property
+    def obstructions(self):
+        return tuple(self._obstructions)
+
+    @property
+    def requirements(self):
+        return tuple(tuple(requirement) for requirement in self._requirements)
+
+    @cssmethodtimer("GriddedPermReduction._minimize_griddedperms")
+    def _minimize_griddedperms(self) -> None:
+        """Minimizes the set of obstructions and the set of requirement lists.
+        The set of obstructions are first reduced to a minimal set. The
+        requirements that contain any obstructions are removed from their
+        respective lists. If any requirement list is empty, then the tiling is
+        empty.
+        """
+
+        def set_empty():
+            self._obstructions = (GriddedPerm.empty_perm(),)
+            self._requirements = tuple()
+
+        i = 0
+        while True:
+            i += 1
+            # Minimize the set of obstructions
+            minimized_obs = self.minimal_obs()
+
+            if minimized_obs and not minimized_obs[0]:
+                set_empty()
+                break
+
+            # Minimize the set of requiriments
+            minimized_requirements = self.minimal_reqs(minimized_obs)
+
+            if minimized_requirements and not minimized_requirements[0]:
+                set_empty()
+                break
+
+            if (
+                self._obstructions == minimized_obs
+                and self._requirements == minimized_requirements
+            ):
+                break
+
+            self._obstructions = minimized_obs
+            self._requirements = minimized_requirements
+        GriddedPermReduction.passes[i] += 1
+
```

Comment:
# if there is a requirement for which every component contains the same factor of
    # obstruction, then that factor can be removed from obstruction

Reply:
# uses requirements to shrink obstructions